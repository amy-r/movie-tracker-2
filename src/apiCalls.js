import { key } from './key';

export const test = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1`;

export const fetchParse = async url => {
  try {
    const response = await fetch(url);
    const parsed = await response.json();
    return parsed;
  } catch (error) {
    return 'fetchParse Err';
  }
};

export const movieCleaner = movies => {
  if (movies === null) {
    return undefined;
  }
  return movies.results.map(movie => {
    const {
      title,
      poster_path,
      id,
      release_date,
      vote_average,
      overview
    } = movie;
    return {
      title,
      poster_path,
      movie_id: id,
      release_date,
      vote_average,
      overview
    };
  });
};

export const logIn = async user => {
  try {
    const response = await fetch('/api/users', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed.data;
    }
  } catch (error) {
    return "Error in logIn";
  }
};

export const createUser = async user => {
  try {
    const response = await fetch('/api/users/new', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return "Error in createUser";
  }
};

export const addFavorite = async (movie) => {
  try {
     const response = await fetch('api/users/favorites/new', { 
      method: 'POST', 
      body: JSON.stringify(movie),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in addFavorite';
  }   
}

export const removeFavorite = async (movie) => {
  const { movie_id, user_id } = movie;
  try {
     const response = await fetch(`api/users/${user_id}/favorites/${movie_id}`, { 
      method: 'DELETE', 
      body: JSON.stringify({ movie_id, user_id }),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if(response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in removeFavorite';
  }   
}

export const fetchAllFavorites = async (userId) => {
  try {
    const response = await fetch(`api/users/${userId}/favorites`)
    const favorites = await response.json()
    return favorites.data
  } catch(error) {
      return 'Error in fetchAllFavorites';
  }
  
}