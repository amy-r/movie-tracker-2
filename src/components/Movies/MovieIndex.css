.movie-index {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-row: 2;
  grid-gap: 10px;
  grid-area: 1/13;
}
