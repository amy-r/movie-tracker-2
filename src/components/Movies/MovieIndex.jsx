import React, { Component } from 'react';
import Card from '../Card/Card';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { addFavorite } from '../../actions/actions';

import './MovieIndex.css';
import PropTypes from 'prop-types';

export class MovieIndex extends Component {

  handleClick = (event) => {
    const { favorites, movies, user, addFavorite } = this.props;
    const user_id = user.id;
    const { id } = event.target;
    if (!favorites) {
      movies.find(movie => movie.movie_id === parseInt(id))
    }

    //add or remove from favorites array 
    //use our actions which are add and remove
    //compare movie_id to favorites array to see which action
    // conditional to decide whether to dispatch ADD or REMOVE
    //CHECK TO SEe if it's in the favorites array
    //if it's in there we remove
    //if it's not we add
    // Add Favorite - /users/favorites/new
    // To save a favorite you must send into the 
    // body: movie_id, user_id and title, poster_path, 
    // release_date, vote_average, overview. 
    // Keep in mind the response only gives the new favorite id
  }

  movieCards = () => {
    const { movies } = this.props;
    return movies.map( movie => {
      return <Card movie={movie}
                   handleClick={this.handleClick}
                   key={movie.movie_id} />;
    });
  };

  render() {
    return <div className="movie-index">{this.movieCards()}</div>;
  }
}

const mapStateToProps = state => ({
  movies: state.movies, 
  favorites: state.favorites, 
  user: state.user
});

const mapDispatchToProps = dispatch => ({
  addFavorite: movie => dispatch(addFavorite(movie)) 
})

MovieIndex.propTypes = {
  movies: PropTypes.array
};

export default withRouter(connect(mapStateToProps, null)(MovieIndex));
