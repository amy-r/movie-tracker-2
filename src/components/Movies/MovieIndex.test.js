import React from 'react';
import { MovieIndex, mapStateToProps, mapDispatchToProps } from './MovieIndex';
import { shallow } from 'enzyme';
import * as api from '../../apiCalls';

describe('MOVIE_INDEX', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<MovieIndex movies={[]} user={{}} favorites={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updateFavorites', () => {
    it('should call api.fetchAllFavorites once', () => {
      const wrapper = shallow(<MovieIndex movies={[]} user={{}} favorites={[]} populateFavorites={jest.fn()}/>);
      api.fetchAllFavorites = jest.fn();
      wrapper.instance().updateFavorites();
      wrapper.update();
      expect(api.fetchAllFavorites).toHaveBeenCalledTimes(1);
    });
  });
  

  describe('MDTP and MSTP', () => {
    it('should define movies props for the container MSTP', () => {
      const title = 'The Rocky';
      const poster = '/RockyIsA-BA.jpg';
      const mockStore = { movies: [{ title, poster }] };
      const expected = [{ title, poster }];
      const mapped = mapStateToProps(mockStore);
      expect(mapped.movies).toEqual(expected);
    });

    it('should define user props for the container MSTP', () => {
      const id = 0;
      const name = 'jeffIsGreat';
      const mockStore = { user: [{ id, name }] };
      const expected = [{ id: 0, name: 'jeffIsGreat' }];
      const mapped = mapStateToProps(mockStore);
      expect(mapped.user).toEqual(expected);
    });

    it('should define favorites props for the container MSTP', () => {
      const title = 'The Rocky';
      const poster = '/RockyIsA-BA.jpg';
      const mockStore = { favorites: [{ title, poster }] };
      const expected = [{ title, poster }];
      const mapped = mapStateToProps(mockStore);
      expect(mapped.favorites).toEqual(expected);
    });

    it('should call the dispatch function on MDTP', () => {
      const mockDispatch = jest.fn();
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.populateFavorites();
      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});
