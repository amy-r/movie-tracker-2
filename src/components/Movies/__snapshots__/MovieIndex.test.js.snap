// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MOVIE_INDEX should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieIndex
    favorites={Array []}
    movies={Array []}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "movie-index",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "movie-index",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
